src/persist-txn/src/txn_write.rs

```diff
@@ -141,10 +143,24 @@ where
                 txns_updates.push((*data_id, batch_raw));
             }
 
-            let txns_updates = txns_updates
+            let mut txns_updates = txns_updates
                 .iter()
                 .map(|(k, v)| ((k, v), &commit_ts, 1))
                 .collect::<Vec<_>>();
+            let apply_is_empty = txns_updates.is_empty();
+
+            // Tidy guarantees that anything in retractions has been applied,
+            // but races mean someone else may have written the retraction. If
+            // the following CaA goes through, then the `update_gt(txns_upper)`
```

Comment:
And confirmed! This was tricky to manifest in a test because the `txns::recent_upper` call at the beginning of `commit_at` meant we needed true concurrency. I actually think it's a good idea to remove that call, anyway, since it's a perf improvement when pubsub is missing (maelstrom) but a perf pessimization when pubsub is present (ahem... mz prod)

I'll push an update with the fix and the test tomorrow. I'm a little bummed that the conflicting_writes test doesn't seem to catch this under `cargo stress` (I can't think of any reason it wouldn't, especially without the `recent_upper` call), but it doesn't seem to. C'est la vie

Fantastic catch!

Reply:
Aha! conflicting_writes didn't catch it because all those operations incremented the timestamp by 1, so if there was a conflict, it bounced out of the commit_at loop before trying a second time. adding a random value to each commit timestamp repros this bug immediately. I'll add that too